### PR TITLE
Optimize arbos code

### PR DESCRIPTION
**Pull request by edfelten**
Rewrite small sections of ArbOS code to speed up bottlenecks identified using the profiler.

### CODE REVIEW

File: stdlib/bytestream.mini [hunk 1]
```diff
@@ -105,6 +105,21 @@ public func bytestream_get256(bs: ByteStream) -> option<(ByteStream, uint)> {
...
+    // read part of the next word -- just get the first nbytes bytes of it
+    // nbytes must be <= 32
+    return xif (nbytes > 32) {
+        None<(ByteStream, uint)>
+    } elseif (bs.currentOffset+nbytes > bs.capacity) {
+        None<(ByteStream, uint)>
+    } else {
+        Some((
+            bs with { currentOffset: bs.currentOffset+nbytes },
+            asm(256-8*nbytes, getbuffer256(bs.buf, bs.currentOffset+bs.slice)) uint { shr },
```
Comment (hkalodner) [inline stdlib/bytestream.mini:118]: Would be be worth actually caching the `getbuffer256` result when we're reading one by at a time and then reading out of the cache?

File: arb_os/codeSegment.mini [hunk 2]
```diff
@@ -116,15 +117,22 @@ impure func parseEvmCodeSegment(bs: ByteStream) -> EvmInstructionStack {
...
+                } else {
+                    let val = 0;
+                    while (bytesRemaining > 0) {
+                        if let Some(res) = bytestream_getByte(bs) {
+                            let (ubs, v) = res;
+                            bs = ubs;
+                            val = 256*val + v;
+                        }
+                        bytesRemaining = bytesRemaining-1;
                     }
```
Comment (hkalodner) [inline arb_os/codeSegment.mini:133]: Is is possible to hit this case?

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: arb_os/codeSegment.mini (hunk 2 above)
Comment (edfelten) [reply arb_os/codeSegment.mini:133]: There is a corner case where that could happen, when the input byte stream ends unexpectedly before getting to the end of the expected immediate value length.  That does happen in practice, when we're parsing non-code data at the end of a file (without realizing that it's not code).

File: stdlib/bytestream.mini (hunk 1 above)
Comment (edfelten) [reply stdlib/bytestream.mini:118]: It's not clear whether that would be faster. That could involve more procedure calls, and has non-trivial performance tradeoffs when the reads are not 32-byte aligned.

Comment (hkalodner) [reply stdlib/bytestream.mini:118]: Sounds good

File: arb_os/codeSegment.mini (hunk 2 above)
Comment (hkalodner) [reply arb_os/codeSegment.mini:133]: 👍

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**